File: gradle-plugin/karakum/src/jsMain/kotlin/io/github/sgrishchenko/karakum/extension/extractTypeParameters.ts
import ts, {ConditionalTypeNode, Declaration} from "typescript";
import {traverse} from "../utils/traverse.js";
import {TypeScriptService, typeScriptServiceKey} from "./plugins/TypeScriptPlugin.js";
import {ConverterContext} from "./context.js";
import {Render} from "./Render.kt";

export type TypeParameterExtractionResult = [ts.Node, Declaration][]

export function extractTypeParameters(
    node: ts.Node,
    context: ConverterContext
): TypeParameterExtractionResult {
    const typeScriptService = context.lookupService<TypeScriptService>(typeScriptServiceKey)

    const result: TypeParameterExtractionResult = []

    const typeChecker = typeScriptService?.program.getTypeChecker()

    const containerNode = node

    traverse(containerNode, node => {
        if (ts.isIdentifier(node)) {
            const symbol = typeChecker?.getSymbolAtLocation(node)
            const typeParameterDeclarations = (symbol?.declarations ?? [])
                .filter(declaration => ts.isTypeParameterDeclaration(declaration))

            for (const declaration of typeParameterDeclarations) {
                let typeParameterContainer = typeScriptService?.getParent(declaration)

                if (typeParameterContainer && ts.isInferTypeNode(typeParameterContainer)) {
                    const conditionalType = typeScriptService?.findClosest(
                        typeParameterContainer,
                        (node): node is ConditionalTypeNode => ts.isConditionalTypeNode(node)
                    )

                    if (conditionalType != undefined) {
                        typeParameterContainer = conditionalType.trueType
                    }
                }

                const foundParent = typeScriptService?.findClosest(containerNode, node => node === typeParameterContainer)

                if (foundParent !== undefined && foundParent !== containerNode) {
                    result.push([node, declaration])
                }
            }
        }
    })

    return result
}

export function renderDeclaration(result: TypeParameterExtractionResult, render: Render): string {
    return result
        .map(([, declaration]) => render(declaration))
        .filter(Boolean)
        .join(", ")
}

export function renderReference(result: TypeParameterExtractionResult, render: Render): string {
    return result
        .map(([node]) => render(node))
        .filter(Boolean)
        .join(", ")
}

File: gradle-plugin/karakum/src/jsMain/kotlin/io/github/sgrishchenko/karakum/extension/extractTypeParameters.kt
package io.github.sgrishchenko.karakum.extension

import io.github.sgrishchenko.karakum.extension.plugins.TypeScriptService
import io.github.sgrishchenko.karakum.extension.plugins.typeScriptServiceKey
import io.github.sgrishchenko.karakum.util.traverse
import js.array.ReadonlyArray
import typescript.*

typealias TypeParameterExtractionResult = ReadonlyArray<Pair<Node, Declaration>>

fun extractTypeParameters(
    node: Node,
    context: ConverterContext
): TypeParameterExtractionResult {
    val typeScriptService = context.lookupService<TypeScriptService>(typeScriptServiceKey)

    val result = mutableListOf<Pair<Node, Declaration>>()

    val typeChecker = typeScriptService?.program?.getTypeChecker()

    traverse(node) { currentNode ->
        if (isIdentifier(currentNode)) {
            val symbol = typeChecker?.getSymbolAtLocation(currentNode)
            val typeParameterDeclarations = (symbol?.declarations ?: emptyArray())
                .filter { declaration -> isTypeParameterDeclaration(declaration) }

            for (declaration in typeParameterDeclarations) {
                var typeParameterContainer = typeScriptService?.getParent(declaration)

                if (typeParameterContainer != null && isInferTypeNode(typeParameterContainer)) {
                    val conditionalType = typeScriptService?.findClosest(typeParameterContainer) {
                        isConditionalTypeNode(it)
                    }

                    if (conditionalType != null && isConditionalTypeNode(conditionalType)) {
                        typeParameterContainer = conditionalType.trueType
                    }
                }

                val foundParent = typeScriptService?.findClosest(node) { it == typeParameterContainer }

                if (foundParent != null && foundParent != node) {
                    result += currentNode to declaration
                }
            }
        }
    }

    return result.toTypedArray()
}

fun renderDeclaration(result: TypeParameterExtractionResult, render: Render<Node>): String {
    return result
        .map { (_, declaration) -> render(declaration) }
        .filter { it.isNotEmpty() }
        .joinToString(separator = ", ")
}

fun renderReference(result: TypeParameterExtractionResult, render: Render<Node>): String {
    return result
        .map { (node) -> render(node) }
        .filter { it.isNotEmpty() }
        .joinToString(separator = ", ")
}
